app/src/main/kotlin/com/wire/android/ui/home/conversations/model/MessageTypes.kt

```diff
@@ -90,8 +102,70 @@ internal fun MessageBody(
         StrikethroughExtension.builder().requireTwoTildes(true).build(),
         TablesExtension.create()
     )
+    text?.also {
+        MarkdownDocument(Parser.builder().extensions(extensions).build().parse(it) as Document, nodeData)
+    }
+    buttonList?.also {
+        MessageButtonsContent(
+            messageId = messageId,
+            buttonList = it,
+            onClick = onButtonClick,
+            pendingButton = pendingButton
+        )
+    }
+}
 
-    MarkdownDocument(Parser.builder().extensions(extensions).build().parse(text) as Document, nodeData)
+@Composable
+fun MessageButtonsContent(
+    messageId: String,
+    pendingButton: String?,
+    buttonList: List<MessageButton>,
+    onClick: ((messageId: String, buttonId: String) -> Unit)?
+) {
+    Column(
+        modifier = Modifier
+            .wrapContentSize()
+    ) {
+
+        for (index in buttonList.indices) {
+            val button = buttonList[index]
+            MessageButton(
+                messageId = messageId,
+                isPending = pendingButton == button.id,
+                button = button,
+                onClick = onClick
+            )
+            if (index != buttonList.lastIndex) {
+                Spacer(modifier = Modifier.padding(top = dimensions().spacing8x))
+            }
+        }
+    }
+}
+
+@SuppressLint("RememberReturnType")
+@Composable
+fun MessageButton(
+    messageId: String,
+    button: MessageButton,
+    isPending: Boolean,
+    onClick: ((messageId: String, buttonId: String) -> Unit)?
+) {
```

Comment:
This could be a good place where the ViewModel would live

Reply:
i get you idea but this viewModel is supossed to be global for all messages so i will still need to pass lambda down down the path to message body 

tho i will play around with this idea in a following PR and try to to make a ViewModel for single message and see how it goes